onnx/reference/ops/op_affine_grid.py

```diff
@@ -0,0 +1,90 @@
+    size_zeros = np.zeros(data_size)
+    original_grid = [np.ones(data_size)]
+    for dim, dim_size in enumerate(data_size):
+        assert dim_size > 0
+        if align_corners == 1:
+            step = 2.0 / (dim_size - 1)
+            start = -1
+            stop = 1 + 0.0001
+            a = np.arange(start, stop, step)
+        else:
+            step = 2.0 / dim_size
+            start = -1 + step / 2
+            stop = 1
+            a = np.arange(start, stop, step)
+        if dim == 0:
+            if is_2d:
+                y = np.reshape(a, (dim_size, 1)) + size_zeros
+                original_grid = [y, *original_grid]
+            else:
+                z = np.reshape(a, (dim_size, 1, 1)) + size_zeros
+                original_grid = [z, *original_grid]
+        elif dim == 1:
+            if is_2d:
+                x = np.reshape(a, (1, dim_size)) + size_zeros
+                original_grid = [x, *original_grid]
+            else:
+                y = np.reshape(a, (1, dim_size, 1)) + size_zeros
+                original_grid = [y, *original_grid]
+        else:
+            x = np.reshape(a, (1, dim_size)) + size_zeros
+            original_grid = [x, *original_grid]
+    return np.stack(original_grid, axis=2 if is_2d else 3)
+
+
+def apply_affine_transform(theta_n, original_grid_homo):
+    # theta_n: (N, 2, 3) for 2D, (N, 3, 4) for 3D
+    # original_grid_homo: (H, W, 3) for 2D, (D, H, W, 4) for 3D
+    assert theta_n.ndim == 3
+    if original_grid_homo.ndim == 3:
+        N, dim_2d, dim_homo = theta_n.shape
+        assert dim_2d == 2 and dim_homo == 3
+        H, W, dim_homo = original_grid_homo.shape
+        assert dim_homo == 3
+        # reshape to [H * W, dim_homo] and then transpose to [dim_homo, H * W]
+        original_grid_transposed = np.transpose(
+            np.reshape(original_grid_homo, (H * W, dim_homo))
+        )
+        grid_n = np.matmul(
+            theta_n, original_grid_transposed
+        )  # shape (N, dim_2d, H * W)
+        # transpose to (N, H * W, dim_2d) and then reshape to (N, H, W, dim_2d)
+        grid = np.reshape(np.transpose(grid_n, (0, 2, 1)), (N, H, W, dim_2d))
+        return grid
+    else:
+        assert original_grid_homo.ndim == 4
+        N, dim_3d, dim_homo = theta_n.shape
+        assert dim_3d == 3 and dim_homo == 4
+        D, H, W, dim_homo = original_grid_homo.shape
+        assert dim_homo == 4
+        # reshape to [D * H * W, dim_homo] and then transpose to [dim_homo, D * H * W]
+        original_grid_transposed = np.transpose(
+            np.reshape(original_grid_homo, (D * H * W, dim_homo))
+        )
+        grid_n = np.matmul(
+            theta_n, original_grid_transposed
+        )  # shape (N, dim_3d, D * H * W)
+        # transpose to (N, D * H * W, dim_3d) and then reshape to (N, D, H, W, dim_3d)
+        grid = np.reshape(np.transpose(grid_n, (0, 2, 1)), (N, D, H, W, dim_3d))
+        return grid
+
+
+class AffineGrid(OpRun):

```

Comment:
Question: given that affine-grid is a function-op, should we just rely on that for the implementation? or, have a separate implementation? 

What are the pros/cons? Using the function, eliminates duplicate definition, and will test it better. Just thinking aloud, since we will face a similar question for other complex function-ops too ... may be good to have a general guideline.

Reply:
good point! I updated the code to use reference implementation which in turn take advantage of the function body of the body definition. However, I am not sure if not having numpy reference implementation is desirable. For example, people may want to see the sample implementation to understand the op.

Reply:
If this op has a custom implementation, it also means the function body is not tested against the test cases.

Reply:
This is a common issue with function ops: only runtime can validate the function body. For this reason I added test cases for both cases: use or not use reference evaluator.

Reply:
Here is further investigation:
If we do not provide the op's reference implementation AffineGrid(OpRun) - that is: it does not exist or is not registered with ReferenceEvaluator (from onnx.reference.ops.op_affine_grid import AffineGrid in _op_list.py). ReferenceEvaluator can be used to generate expected outputs. In this case the op's function body is used in computation.

If AffineGrid(OpRun) is provided and registered, ReferenceEvaluator can still be used to generated the expected outputs. In this case AffineGrid(OpRun), instead of the op's function body, is used in computation. In this case, however, I would NOT use ReferenceEvaluator because it hides the implementation and defeat the purpose of providing example code. I would bring code from AffineGrid._run() directly into test case generating code.

Back to @gramalingam's comment, I would argue that it is always desirable to have an example/reference implementation to show the computation of an op. We shall always provide an implementation in the form of python methods and call it from test case generating code and reference implementations derived from OpRun. I am very worried on using function body to generate test cases. It will confuse people if the op's function body is incorrect.

I will reimplement the test case generating code to remove uses of ReferenceEvaluator once we conclude this discussion.

Reply:
I have reimplemented the test case generating code to remove uses of ReferenceEvaluator.